BootloaderCorePkg/Library/AcpiInitLib/Fpdt.c

```diff
@@ -0,0 +1,233 @@
+
+  @retval S3 performance table address     Value 0 means not found.
+**/
+UINTN
+GetFpdtS3Table (
+  IN  UINT32                                   AcpiTableBase
+  )
+{
+  EFI_ACPI_5_0_ROOT_SYSTEM_DESCRIPTION_POINTER *Rsdp;
+  EFI_ACPI_DESCRIPTION_HEADER                  *Rsdt;
+  EFI_ACPI_COMMON_HEADER                       *Hdr;
+  UINT32                                       *RsdtEntry;
+  UINT32                                       NumEntries;
+  FIRMWARE_PERFORMANCE_TABLE                   *Fpdt;
+  UINT8                                        Index;
+
+  Rsdp = (EFI_ACPI_5_0_ROOT_SYSTEM_DESCRIPTION_POINTER *) AcpiTableBase;
+  Rsdt = (EFI_ACPI_DESCRIPTION_HEADER *) (UINTN) Rsdp->RsdtAddress;
+
+  NumEntries = (Rsdt->Length - sizeof(EFI_ACPI_DESCRIPTION_HEADER)) / sizeof(UINT32);
+  RsdtEntry  = (UINT32 *) ((UINT8 *)Rsdt + sizeof (EFI_ACPI_DESCRIPTION_HEADER));
+  for (Index = 0; Index < NumEntries; Index++) {
+    Hdr = (EFI_ACPI_COMMON_HEADER *) (UINTN) RsdtEntry[Index];
+    if (Hdr->Signature == EFI_ACPI_5_0_FIRMWARE_PERFORMANCE_DATA_TABLE_SIGNATURE) {
+      Fpdt = (FIRMWARE_PERFORMANCE_TABLE *) Hdr;
+      if (DEBUG_PRINT_FPDT_DATA == 1) {
+        BOOT_PERFORMANCE_TABLE  *BootTable;
+        BootTable = (BOOT_PERFORMANCE_TABLE *)(UINTN)Fpdt->BootPointerRecord.BootPerformanceTablePointer;
+        DEBUG ((EFI_D_INFO, "FPDT: ResetEnd                = %ld\n", BootTable->BasicBoot.ResetEnd));
+        DEBUG ((EFI_D_INFO, "FPDT: OsLoaderLoadImageStart  = %ld\n", BootTable->BasicBoot.OsLoaderLoadImageStart));
+        DEBUG ((EFI_D_INFO, "FPDT: OsLoaderStartImageStart = %ld\n", BootTable->BasicBoot.OsLoaderStartImageStart));
+        DEBUG ((EFI_D_INFO, "FPDT: ExitBootServicesEntry   = %ld\n", BootTable->BasicBoot.ExitBootServicesEntry));
+        DEBUG ((EFI_D_INFO, "FPDT: ExitBootServicesExit    = %ld\n", BootTable->BasicBoot.ExitBootServicesExit));
+      }
+      return (UINTN)Fpdt->S3PointerRecord.S3PerformanceTablePointer;
+    }
+  }
+
+  return 0;
+}
+
+
+/**
+  Update ACPI FPDT S3 performance record table.
+
+  @param[in] AcpiTableBase     ACPI base address
+
+  @retval EFI_SUCCESS          Update the S3 performance table successfully.
+  @retval Others               Failed to update the table.
+ **/
+EFI_STATUS
+EFIAPI
+UpdateFpdtS3Table (
+  IN  UINT32                          AcpiTableBase
+  )
+{
+  S3_PERFORMANCE_TABLE                *S3PerfTable;
+  BL_PERF_DATA                        *PerfData;
+  UINT64                              TscValue;
+  UINT64                              TimeInMs;
+  UINT64                              TotalResumeTime;
+  EFI_ACPI_5_0_FPDT_S3_RESUME_RECORD  *S3Resume;
+
+  S3PerfTable = (S3_PERFORMANCE_TABLE *)GetFpdtS3Table (AcpiTableBase);
+  if (S3PerfTable == NULL) {
+    return EFI_NOT_FOUND;
+  }
+
+  ASSERT (S3PerfTable->Header.Signature == EFI_ACPI_5_0_FPDT_S3_PERFORMANCE_TABLE_SIGNATURE);
+
+  // Get last measured time
+  PerfData = GetPerfDataPtr();
```

Comment:
We should read current TSC instead of the last entry in the performance table.

Reply:
agree. will update code.